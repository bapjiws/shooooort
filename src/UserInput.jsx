import React, { Component } from 'react';

import { Row, Col, Button, FormControl } from 'react-bootstrap/lib';

import { shortenLink, getShortcodeStats } from './actions/shortcode';

export default class Headline extends Component {
    constructor(props) {
        super(props);
        this.state = {
            input: ''
        };
    }

    handleChange = event => {
        console.log('event:', event.target.value);
        this.setState({input: event.target.value});
    };

    render() {
        return <Row>
            <Col xs={2} md={2}></Col>
            <Col xs={6} md={6}>
                <form>
                    <FormControl
                        className={this.state.input ? "form-with-input" : "form-no-input"}
                        type="text"
                        value={this.state.input}
                        placeholder="Paste the link you want to shorten here"
                        onChange={this.handleChange}
                    />
                </form>
            </Col>
            <Col xs={2} md={2}>
                <Button
                    className={this.state.input ? "button-with-input" : "button-no-input"}
                    disabled={this.state.input === ''}
                    //onClick={() => getShortcodeStats(this.state.input)}
                    onClick={() => shortenLink(this.state.input)}
                >
                    Shorten this link
                </Button></Col>
            <Col xs={2} md={2}></Col>
        </Row>
    }
}