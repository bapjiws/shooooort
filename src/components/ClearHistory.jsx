import React from 'react';
import { connect } from 'react-redux';

import { Row, Col } from 'react-bootstrap/lib';

import { clearHistory } from '../../redux/actions/linksData';

const ClearHistory = ({ clearHistory }) =>
    <Row className="clear-history flex-cross-axis-align-baseline">
        <Col className="text-section-heading" md={6}>
            Previously shortened by you
        </Col>
        <Col md={2}>
            <a className="text-clear-history-link" onClick={clearHistory}>Clear history</a>
        </Col>
    </Row>;

export default connect(
    null,
    { clearHistory }
)(ClearHistory);