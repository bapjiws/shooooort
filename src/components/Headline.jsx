import React from 'react';

const Headline = () => (
    <header className="headline-container">
        <div className="logo-container text-logo">URL&nbsp;Shortener</div>
        <div className="description-container text-normal">Does exactly what it's supposed to.</div>
    </header>
);

export default Headline;