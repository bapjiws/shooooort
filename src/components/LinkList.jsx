import React, { Component } from 'react';
import { connect } from 'react-redux';

import TimeAgo from 'react-timeago';
import { Row, Col } from 'react-bootstrap/lib';
import CopyToClipboard from 'react-copy-to-clipboard';

import { fetchLinksInfo } from '../../redux/actions/linksData';

class LinkList extends Component {
    componentDidMount() {
        this.props.fetchLinksInfo();
        this.timerID = setInterval(
            () => this.props.fetchLinksInfo(),
            1000
        );
    }

    componentWillUnmount() {
        clearInterval(this.timerID);
    }

    render() {
        // TODO: ideally, should only re-render if the incoming data is different -- reselect could be an option.

        const { linksData } = this.props;
        return <div>
            <Row className="link-list-header">
                <Col className="text-table-heading" md={8}>
                    LINK
                </Col>
                <Col className="text-table-heading text-align-center" md={1}>
                    VISITS
                </Col>
                <Col className="text-table-heading text-align-center" md={3}>
                    LAST VISITED
                </Col>
            </Row>
            {
                Object.keys(linksData).map(key =>
                    <Row className="link-list-row flex-cross-axis-align-center" key={key}>
                        <Col md={8}>
                            <CopyToClipboard text={`shooooort.com/${key}`}>
                                <div><span className="text-url-body">shooooort.com/</span><span className="text-url-shortcode">{ `${key}` }</span></div>
                            </ CopyToClipboard>

                            <div className="text-url-original">{ linksData[key].url }</div>
                        </Col>
                        <Col className="text-data text-align-center" md={1}>
                            { linksData[key].redirectCount }
                        </Col>
                        <Col className="text-data text-align-center" md={3}>
                            { <TimeAgo date={linksData[key].lastSeenDate} live={false} /> }
                        </Col>
                    </Row>
                )
            }
        </div>;
    }
}

export default connect(
    state => ({
        linksData: state.linksData.data
    }),
    { fetchLinksInfo }
)(LinkList);