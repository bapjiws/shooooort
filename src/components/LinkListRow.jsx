import React, { Component } from 'react';

import TimeAgo from 'react-timeago';
import CopyToClipboard from 'react-copy-to-clipboard';

class LinkListRow extends Component {
    constructor(props) {
        super(props);

        this.state = {
            mouseOver: false,
            clicked: false
        };
    }

    render() {
        const { rowId, linksData } = this.props;
        const { mouseOver, clicked } = this.state;

        return (
            <div className="link-list-row-container" key={rowId}>
                <div className="link-list-row-link-container">
                    <CopyToClipboard text={`http://goo.gl/${rowId}`}>
                        <div>
                            <span
                                onMouseOver={() => this.setState({mouseOver: true})}
                                onMouseLeave={() => {
                                    this.setState({mouseOver: false});
                                    this.setState({clicked: false})
                                }}
                                onClick={() => this.setState({clicked: true})}
                            >
                                <span className="text-url-body">http://goo.gl/</span>
                                <span className="text-url-shortcode">{ `${rowId}` }</span>
                            </span>
                            <span
                                className={mouseOver && !clicked ?
                                    'copy-suggestion-visible' : 'copy-suggestion-hidden'}>
                                            Click to copy this link
                                        </span>
                        </div>
                    </ CopyToClipboard>
                    <a className="text-url-original" href={ linksData[rowId].url } target="_blank">{ linksData[rowId].url }</a>
                </div>
                <div className="link-list-row-visits-container text-data">
                    { linksData[rowId].visits }
                </div>
                <div className="link-list-row-last-visited-container text-data">
                    { <TimeAgo date={linksData[rowId].lastVisited} live={false} /> }
                </div>
            </div>
        )
    }
}

export default LinkListRow;