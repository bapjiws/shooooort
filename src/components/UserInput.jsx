import React, { Component } from 'react';
import { connect } from 'react-redux';

import { Button, FormGroup, FormControl, HelpBlock } from 'react-bootstrap/lib';

import { shortenLink } from '../../redux/actions/linksData';

import { inputIsValid } from '../../utils/inputValidation';

class UserInput extends Component {
    constructor(props) {
        super(props);
        this.state = {
            input: '',
            inputIsValid: false
        };
        this.form = null;
    }

    handleChange = event => {
        const input = event.target.value;
        this.setState({
            input,
            inputIsValid: inputIsValid(input)
        });
    };

    handleCLick = () => {
        this.props.shortenLink(this.state.input);
        this.setState({input: ''});
    };

    handleKeyDown = event => {
        switch (event.key) {
            case 'Enter':
                if (!(this.state.input && this.state.inputIsValid)) {
                    event.preventDefault();
                    return;
                }

                event.preventDefault();
                this.props.shortenLink(this.state.input);
                this.setState({input: ''});
                this.form.blur();
                break;
        }
    };

    getValidationState = () => {
        const input = this.state.input;
        if (!input) return;
        return inputIsValid(input) ? 'success' : 'error';
    };

    render() {
        const { input, inputIsValid } = this.state;

        return <div className="user-input">
            <form className="form">
                <FormGroup
                    validationState={this.getValidationState()}
                >
                    <FormControl
                        className="text-form-control"
                        type="text"
                        value={this.state.input}
                        placeholder="Paste the link you want to shorten here"
                        onChange={this.handleChange}
                        onKeyDown={this.handleKeyDown}
                        inputRef={ref => this.form = ref}
                    />
                    <FormControl.Feedback />
                    { input && !inputIsValid && <HelpBlock className="text-form-control-validation-help">Links should start with "http://" or "https://"</HelpBlock> }
                </FormGroup>
            </form>
            <div className="button">
                <Button
                    disabled={!input || !inputIsValid}
                    onClick={this.handleCLick}
                >
                    Shorten this link
                </Button>
            </div>
        </div>
    }
}

export default connect(
    null, // mapStateToProps
    { shortenLink }
)(UserInput);