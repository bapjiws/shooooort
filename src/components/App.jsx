import React from 'react';

import '../../styles/main';

import { Grid } from 'react-bootstrap/lib';

import Headline from './Headline';
import UserInput from './UserInput';
import ClearHistory from './ClearHistory';
import LinkList from './LinkList';

const App = () => {
    return <Grid>
        <Headline/>
        <UserInput/>
        <ClearHistory/>
        <LinkList/>
    </Grid>
};

export default App;