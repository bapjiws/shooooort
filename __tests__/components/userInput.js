import React from 'react';
import { shallow, mount } from 'enzyme';
import toJson from 'enzyme-to-json'; // Documentation: https://github.com/adriantoine/enzyme-to-json

// Import the unconnected version -- see http://redux.js.org/docs/recipes/WritingTests.html#connected-components
import { UserInput } from '../../src/components/UserInput';

let wrapper, inputHelp;
beforeEach(() => {
    wrapper = shallow(<UserInput />);
    inputHelp = <div className="text-validation-help input-help">Links should start with "http://" or "https://"</div>;
});

describe('UserInput', () => {
    it('renders correctly', () => {
        expect(toJson(wrapper)).toMatchSnapshot();
    });

    it('should show feedback, cross icon and have "incorrect" style on focus w/ invalid input', () => {
        // const { wrapper, inputHelp } = t.context.data;

        expect(wrapper.find('input[type="text"]').hasClass('text-input')).toBe(true);

        'http:/'.split('').reduce((acc, curr) => {
            acc += curr;
            wrapper.find('input[type="text"]').simulate('change', { target: { value: acc } });

            expect(wrapper.state('input')).toBe(acc);
            expect(
                wrapper.find('input[type="text"]').hasClass('text-input input-focus-incorrect')
            ).toBe(true);
            expect(wrapper.contains(inputHelp)).toBe(true);
            expect(wrapper.find('svg').hasClass('incorrect-input-icon')).toBe(true);
            return acc;
        }, '');
    });

    it('should not show feedback, contain check icon and have "correct" style on focus w/ valid input', () => {
        // const { wrapper, inputHelp } = t.context.data;

        wrapper.find('input[type="text"]').simulate('change', { target: { value: 'http://' } });

        expect(
            wrapper.find('input[type="text"]').hasClass('text-input input-focus-correct')
        ).toBe(true); // or simply input-focus-correct
        expect(wrapper.contains(inputHelp)).toBe(false);
        expect(wrapper.find('svg').hasClass('correct-input-icon')).toBe(true);
    });

    it('should blur, fire shortenLink and clear the input field on Enter w/ valid input', () => {
        // It seems that wrapper.ref() does not work with the new "ref" style (see https://github.com/airbnb/enzyme/issues/298
        // and https://github.com/airbnb/enzyme/issues/566. We will mount the component into the document and detect
        // changes in document.activeElement.
        document.body.innerHTML = '<div id="root"></div>';
        const shortenLink = jest.fn();
        const wrapper = mount(<UserInput shortenLink={shortenLink}/>, { attachTo: document.body.firstChild });

        // Put the cursor into the input field. The first three lines are there for the sake of exercise, because the same
        // won't work with blur for some reason (commented out below). We'll be relying on document.activeElement.
        const focusSpy = jest.spyOn(wrapper.find('input[type="text"]').getDOMNode(), 'focus');
        wrapper.find('input[type="text"]').getDOMNode().focus();
        expect(focusSpy).toHaveBeenCalled();
        expect(document.activeElement.nodeName).toBe('INPUT');

        // Type correct input and hit Enter
        const blurSpy = jest.spyOn(wrapper.find('input[type="text"]').getDOMNode(), 'blur');
        wrapper.find('input[type="text"]').simulate('change', { target: { value: 'http://' } });
        expect(wrapper.state('input')).toBe('http://');
        wrapper.find('input[type="text"]').simulate('keyDown', {key: 'Enter', keyCode: 13});
        expect(wrapper.state('input')).toBe('');
        expect(shortenLink).toHaveBeenCalled(); // shortenLink is tested separately in actions
        expect(blurSpy).toHaveBeenCalled();
        expect(document.activeElement.nodeName).toBe('BODY');


        // Trying to test form submission (needed to test if event.preventDefault() has been called on Enter). Won't work:
        // const submitSpy = jest.spyOn(wrapper.find('form').getDOMNode(), 'submit');
        // wrapper.find('input[type="text"]').simulate('change', { target: { value: 'http://' } });
        // wrapper.find('input[type="text"]').simulate('keyDown', { key: 'Enter', keyCode: 13 });
        // expect(submitSpy).toHaveBeenCalled();
    });
});