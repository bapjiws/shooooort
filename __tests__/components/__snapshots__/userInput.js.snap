// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserInput renders correctly 1`] = `
<div
  className="user-input-container"
>
  <form
    className="form-container"
  >
    <input
      className="text-input"
      onChange={[Function]}
      onKeyDown={[Function]}
      placeholder="Paste the link you want to shorten here"
      type="text"
      value=""
    />
  </form>
  <div
    className="button-container"
  >
    <button
      disabled={true}
      onClick={[Function]}
    >
      Shorten this link
    </button>
  </div>
</div>
`;
